extras/indexing/src/main/java/org/apache/rya/indexing/mongodb/pcj/BasePcjIndexer.java

```diff
@@ -0,0 +1,138 @@
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.groupingBy;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.rya.api.domain.RyaStatement;
+import org.apache.rya.api.domain.RyaURI;
+import org.apache.rya.indexing.entity.model.Entity;
+import org.apache.rya.indexing.entity.storage.EntityStorage;
+import org.apache.rya.indexing.entity.storage.EntityStorage.EntityStorageException;
+import org.apache.rya.indexing.pcj.storage.mongo.MongoPcjDocuments;
+import org.apache.rya.mongodb.MongoDBRdfConfiguration;
+import org.apache.rya.mongodb.MongoSecondaryIndex;
+import org.openrdf.model.URI;
+
+import edu.umd.cs.findbugs.annotations.DefaultAnnotation;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * A base class that may be used to update an {@link EntityStorage} as new
+ * {@link RyaStatement}s are added to/removed from the Rya instance.
+ */
+@DefaultAnnotation(NonNull.class)
+public abstract class BasePcjIndexer implements PcjIndexer, MongoSecondaryIndex {
+
+    private final AtomicReference<MongoDBRdfConfiguration> configuration = new AtomicReference<>();
+    private final AtomicReference<MongoPcjDocuments> pcjDocs = new AtomicReference<>();
+
+    @Override
+    public void setConf(final Configuration conf) {
+        requireNonNull(conf);
+        pcjDocs.set( getPcjStorage(conf) );
+    }
+
+    @Override
+    public Configuration getConf() {
+        return configuration.get();
+    }
+
+    @Override
+    public void storeStatement(final RyaStatement statement) throws IOException {
+        requireNonNull(statement);
+        storeStatements( singleton(statement) );
+    }
+
+    @Override
+    public void storeStatements(final Collection<RyaStatement> statements) throws IOException {
+        requireNonNull(statements);
+
+        final Map<RyaURI,List<RyaStatement>> groupedBySubject = statements.stream()
+            .collect(groupingBy(RyaStatement::getSubject));
+
+        for(final Entry<RyaURI, List<RyaStatement>> entry : groupedBySubject.entrySet()) {
+            try {
+                updateEntity(entry.getKey(), entry.getValue());
+            } catch (final EntityStorageException e) {
+                throw new IOException("Failed to update the Entity index.", e);
+            }
+        }
+    }
+
+    /**
+     * Updates a {@link Entity} to reflect new {@link RyaStatement}s.
+     *
+     * @param subject - The Subject of the {@link Entity} the statements are for. (not null)
+     * @param statements - Statements that the {@link Entity} will be updated with. (not null)
+     */
+    private void updateEntity(final RyaURI subject, final Collection<RyaStatement> statements) throws EntityStorageException {
+        requireNonNull(subject);
```

Comment:
May want to change the name of this method -- Entity doesn't make sense in a pcj context.

Reply:
This wasn't flushed out since this is where the mongo equivalent of fluo would update the PCJ entries.